picard/ui/logview.py

```diff
@@ -97,48 +54,271 @@ def restoreWindowState(self, position, size):
             self.move(pos)
         self.resize(config.persist[size])
 
+    def closeEvent(self, event):
+        return super().closeEvent(event)
+
+
+class LogViewCommon(LogViewDialog):
+    WIDTH = 570
+    HEIGHT = 400
+
+    def __init__(self, log_tail, *args, **kwargs):
+        self.displaying = False
+        self.prev = -1
+        self.log_tail = log_tail
+        self.log_tail.updated.connect(self._updated)
+        super().__init__(*args, **kwargs)
+
+    def show(self):
+        self.display(clear=True)
+        super().show()
+
+    def _updated(self):
+        if self.displaying:
+            return
+        self.display()
+
+    def display(self, clear=False):
+        self.displaying = True
+        if clear:
+            self.prev = -1
+            self.doc.clear()
+            self.textCursor.movePosition(QtGui.QTextCursor.Start)
+        for logitem in self.log_tail.contents(self.prev):
+            self._add_entry(logitem)
+            self.prev = logitem.pos
+        self.displaying = False
+
+    def _add_entry(self, logitem):
+        self.textCursor.movePosition(QtGui.QTextCursor.End)
+        self.textCursor.insertText(logitem.message)
+        self.textCursor.insertBlock()
+        sb = self.browser.verticalScrollBar()
+        sb.setValue(sb.maximum())
+
+
+class Highlighter(QtGui.QSyntaxHighlighter):
+    def __init__(self, string, parent=None):
+        super().__init__(parent)
+
+        self.fmt = QtGui.QTextCharFormat()
+        self.fmt.setBackground(QtCore.Qt.lightGray)
+
+        self.reg = QtCore.QRegExp()
+        self.reg.setPatternSyntax(QtCore.QRegExp.Wildcard)
+        self.reg.setCaseSensitivity(QtCore.Qt.CaseInsensitive)
+        self.reg.setPattern(string)
+
+    def highlightBlock(self, text):
+        expression = self.reg
+        index = expression.indexIn(text)
+        while index >= 0:
+            length = expression.matchedLength()
+            self.setFormat(index, length, self.fmt)
+            index = expression.indexIn(text, index + length)
+
+
+class VerbosityMenu(QtWidgets.QMenu):
+    verbosity_changed = QtCore.pyqtSignal(int)
+
+    def __init__(self, parent=None):
+        super().__init__(parent=parent)
+
+        self.action_group = QtWidgets.QActionGroup(self)
+        self.action_group .setExclusive(True)
+        self.actions = {}
+        for level, feat in log.levels_features.items():
+            act = QtWidgets.QAction(_(feat.name), self, checkable=True)
+            act.triggered.connect(partial(self.verbosity_changed.emit, level))
+            self.action_group.addAction(act)
+            self.addAction(act)
+            self.actions[level] = act
+
+    def set_verbosity(self, level):
+        self.actions[level].setChecked(True)
+
 
 class LogView(LogViewCommon):
```

Comment:
I get this warning from Qt - `QLayout: Attempting to add QLayout "" to LogView "", which already has a layout`
should it matter?

Reply:
That's a parenting conflict, when you set explicitly a parent and Qt think different.
Too much parenting kills parenting ;)
Try to see if it is better with https://github.com/metabrainz/picard/pull/834/commits/5699bb14d1ee81f6053c91c08c18d2f01d0d2982